### PR TITLE
(5) 2365 upload a document:  AzureDocumentRepository, status patch, and JS component 

### PR DESCRIPTION
**Pull request by maiyerlee**
## What changed

TL;DR - create a patch status endpoint; add a temporary testing page and component for uploading/downloading; and update the AzureDocumentRepository.

Database:  
- Add UPDATE_DOCUMENT event to ops event type
- Add new attribute document_size to document table
- Rename file_name to document_name

Backend: 
- Add PATCH document permission to users
- Update document model according to database changes
- Create new UPDATE_DOCUMENT event 
- Remove app registration of document blueprint
- Create new configs for document provider
- API
  - Create a document api class to organize endpoints (similar to existing endpoints)
  - Add document endpoints to OPS list of urls and views 
  - Rename GET documents to /api/v1/documents/{agreement_id}/ and update 
  - Add new endpoint for updating status: /api/v1/documents/{document_id}/status/
  - Fix get documents by agreement id response schema 
  - Update post document request schema to include type, name and size
- Azure Document Repository
  - Add a method to create a reservation into the database when a new document is created
  - Add a method to create an account level Shared Access Signature (SAS) token URL
  - Update the document gateway to use ProviderTypes enums and select provider based on user id
- Fake Document Repository	
  - Add update status functionality, make return on creating document consistent 
- Service
  - Add functionality to query the document table on document_id (uuid) match and update status 
- Document utility
  - Fix usage of user id via flask_jwt_extended.current_id.id
  - Add document type mappings
  - Add database functions to create a new document record, to update a document status after a successful upload, to get all document records by agreement ID

Frontend:
- Add new package "@azure/storage-blob" to application dependencies
- E2E
  - Create test to check that the /upload-document page loads
  - Create test that stores document in memory 
  - Add sample document used in e2e test to cypress fixtures
- Add document endpoints to opsAPI 
- Create a new temporary JS page for testing purposes (UploadDocument.jsx)
- Add functionality to convert file size to MB; validate file extension; process in-memory and azure uploads and downloads; and create storage account JS client 

## Issue

[Upload a document #2365](https://github.com/HHS/OPRE-OPS/issues/2365)

### Demo Videos:
In-memory

https://github.com/user-attachments/assets/41ce1cca-9409-472c-8982-90e59bbb20a9

Azure:

https://github.com/user-attachments/assets/707b919f-450b-4f98-9e55-2319da15970e

## How to test

Test database:
- Verify that the document table contains these columns:
Screenshot 2024-08-12 at 2 30 37 AM

Test in-memory uploads:
1. Launch the application and navigate to the /upload-document page.
2. Ensure that the document table is empty.
3. Select and upload a document (Word, PDF, or Excel) to an agreement associated with the logged-in user.
4. Verify Document Record:
- Confirm that the document table now contains a record.
- Check that this record includes a UUID that matches the UUID logged in the application's console logs.
5. Validate Document Retrieval:
- Use the "Get Documents by Agreement ID" feature to download the document.
- Verify that the downloaded document is the same as the one you uploaded.

Testing azure uploads LOCALLY: 
1. Ensure that the DOCUMENT_STORAGE_ACCOUNT_ACCESS_KEY and DOCUMENT_STORAGE_ACCOUNT_NAME credentials are correctly configured and accessible in the backend system.
2. Follow the steps outlined above for testing in-memory uploads.
3. Confirm that the document has been successfully uploaded by chekcing the DOCUMENT_STORAGE_ACCOUNT_NAME storage account and verifying that the document appears in the correct `documents` container.
 
## Definition of Done Checklist
- [x] OESA: Code refactored for clarity
- [x] OESA: Dependency rules followed
- [x] Automated unit tests updated and passed
- [x] Automated integration tests updated and passed
- [x] Automated quality tests updated and passed
- [x] Automated load tests updated and passed 
- [x] Automated a11y tests updated and passed <del>
- [x] Automated security tests updated and passed
- [x] 90%+ Code coverage achieved
- [x] Form validations updated 

## Links

Links to prior PRs:
1. https://github.com/HHS/OPRE-OPS/pull/2406
4. https://github.com/HHS/OPRE-OPS/pull/2523
5. https://github.com/HHS/OPRE-OPS/pull/2539 
6. https://github.com/HHS/OPRE-OPS/pull/2553
8. https://github.com/HHS/OPRE-OPS/pull/2580

## Improvements 
- Make document status a list of enums (uploading, uploaded, error, etc.)
- Add upper environment storage account configurations to backend environmental variables.

### CODE REVIEW

**Comment by maiyerlee** [comment]
Backend unit tests are failing because the document model import is missing from `backend/models/init.py`. I'll leave the pull request in draft status until [PR#2580](https://github.com/HHS/OPRE-OPS/pull/2580) is merged (since [PR#2580](https://github.com/HHS/OPRE-OPS/pull/2580) contains the import).

File: frontend/src/components/Agreements/Documents/Document.js [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+         // Get a reference to the specific blob (document) using its uuid
+         const blobClient = containerClient.getBlobClient(documentId);
+
+         // Download the blob
+         const downloadBlobResponse = await blobClient.download();
+         const blob = await downloadBlobResponse.blobBody;
+
+         triggerDownload(blob, fileName);
+     } catch (error) {
+         console.error(`Error downloading blob=${documentId}:`, error);
```
[the remark this answers is not in the document]
Comment (maiyerlee) [reply frontend/src/components/Agreements/Documents/Document.js]: I switched to using a fixed format string and separate arguments for variables.

File: frontend/src/components/Agreements/Documents/UploadDocument.jsx [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+        try {
+            // Call the endpoint to get documents by Agreement ID
+            const response = await getDocumentsByAgreementId(getDocumentAgreementId);
+            console.log("getDocumentsByAgreementId response", response);
+
+            const {url, documents} = response;
+
+            if (documents.length > 0) {
+                for (const document of documents) {
+                    console.log(`Downloading document ${document.document_id}`, document);
```
[the remark this answers is not in the document]
Comment (maiyerlee) [reply frontend/src/components/Agreements/Documents/UploadDocument.jsx]: I switched to using a fixed format string and separate arguments for variables.

File: frontend/src/components/Agreements/Documents/Document.js [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    }
+};
+
+// Process file upload based on the specified storage repository
+export const processUploading = async (sasUrl, uuid, file, agreementId, inMemoryUpload, inBlobUpload, patchStatus) => {
+    try {
+        if (sasUrl.includes("FakeDocumentRepository")) {
+            // Upload to an in-memory storage repository
+            await inMemoryUpload(uuid, file);
+        } else if (sasUrl.includes("blob.core.windows.net")) {
```
[the remark this answers is not in the document]
Comment (maiyerlee) [reply frontend/src/components/Agreements/Documents/Document.js]: I created specific allowed host substring check.

File: backend/ops_api/ops/document/document_gateway.py [hunk 4]
```diff
@@ -9,35 +10,39 @@
...
+        self.providers = config.get("AUTHLIB_OAUTH_CLIENTS", {})
         self.repository_factory = DocumentRepositoryFactory()
 
         # Validate and register providers with the factory
-        self.register_provider(self.provider)
+        self.register_providers(self.providers)
 
-    def register_provider(self, provider: str) -> None:
+        # Select provider based on the current user
+        if current_user.id >= 500:  # Users with id 5xx are test users
```
Comment (johndeange) [inline backend/ops_api/ops/document/document_gateway.py:20]: As per our demo, please replace this with using the `config` object (file based config)

**Review by johndeange — APPROVED** [review]

**Comment by maiyerlee** [comment]
Feedback from review session with @johndeange:
- Use opsapi (RTK query) instead of ApplicationContext
- Add new config to determine "document" provider

**Review by fpigeonjr — APPROVED** [review]
⭐ tested locally and manual tests work great
